tools/socktap/cohda.cpp

```diff
@@ -1,41 +1,149 @@
 #include "cohda.hpp"
-#include <vanetza/dcc/data_request.hpp>
-#include <cassert>
-#include <mk2mac-api-types.h>
+#include <boost/crc.hpp>
+#include <iostream>
+#include <llc-api.h>
+#include <vanetza/geonet/router.hpp>
+#include <vanetza/geonet/address.hpp>
+#include <vanetza/dcc/profile.hpp>
+#include <vanetza/common/byte_order.hpp>
 
-namespace vanetza
-{
+namespace vanetza {
 
-using dcc::Profile;
+struct LLCHeader {
+    uint8_t DSAP = 0xAA;
+    uint8_t SSAP = 0xAA;
+    uint8_t ControlField = 0x03;
+    uint8_t OUI[3] = { 0x00, 0x00, 0x00 };
+    uint16_t ProtocolID = geonet::ether_type.net();
+};
 
-ByteBuffer create_cohda_tx_header(const dcc::DataRequest& request)
-{
-    ByteBuffer header(sizeof(tMK2TxDescriptor), 0x00);
-    assert(header.size() == sizeof(tMK2TxDescriptor));
-    tMK2TxDescriptor* tx = reinterpret_cast<tMK2TxDescriptor*>(header.data());
-    tx->ChannelNumber = 180;
+/* See ETSI EN 302 663 V1.2.1 (2013-07), Table B.3 */
+enum tPriority {
+    AC_BK = 1, AC_BE = 3, AC_VI = 5, AC_VO = 6
+};
+constexpr uint8_t QOS_PRIO_MASK = 0x07;
+
+/**
+ * 1 bit payload type (MSDU = 0)
+ * 2 bits ack policy (No_Ack = 1)
+ * 1 bit EOSP (service period = 0)
+ * 1 bit unused
+ * 3 bits prioriy (set dynamically according to traffic class)
+ * 8 bits TXOP (shall be set to 0 when communicating outside of a BSS)
+ */
+struct tQOSControl {
+    uint8_t QOSFlags = 0x20;
+    uint8_t TXOP = 0; // not used
+};
+
+struct tIEEE80211PHeader {
+    uint8_t FrameControlVersionAndType = 0x88;
+    uint8_t FrameControlFlags = 0x00; // not used
+    uint16_t DurationID = 0x00; // not used
+    uint8_t Destination[6];
+    uint8_t Source[6];
+    uint8_t BSSID[6] = { 0xff, 0xff, 0xff, 0xff, 0xff, 0xff }; // Operation outside of a BSS, (see ETSI EN 302 663 V1.2.1 Annex A)
+    uint16_t SequenceControl = 0x00; // 12 bits for sequence number, 4 bits for fragment number; will be set by Cohda LLC
+    tQOSControl QOSControl;
+};
```

Comment:
We might want to add this to the `net` directory where also our `EthernetHeader` is located.